setup.py

```diff
@@ -29,6 +29,11 @@
     'async_generator',
     'boto3',
 
+    # pyopenssl depends on a newer version of cryptography since 22.1.0
+    # TypeError: deprecated() got an unexpected keyword argument 'name'
+    # https://github.com/pyca/pyopenssl/commit/a145fc3bc6d2e943434beb2f04bbf9b18930296f
+    'pyopenssl<22.1.0',
```

Comment:
boto3 is pulling in pyopenssl, this pinning should be before boto3 to make sure the pyopenssl<22.1.0 is installed before boto request pyopenssl.

Reply:
This is an extract of the stacktrace, it fails at "import boto3"

```
2022-09-29 11:21:55.772667 | focal-medium |   File "/home/ubuntu/src/review.opendev.org/openstack/charm-cinder/.tox/func-target/lib/python3.8/site-packages/zaza/openstack/charm_tests/glance/setup.py", line 20, in <module>
2022-09-29 11:21:55.772675 | focal-medium |     import boto3
2022-09-29 11:21:55.772681 | focal-medium |   File "/home/ubuntu/src/review.opendev.org/openstack/charm-cinder/.tox/func-target/lib/python3.8/site-packages/boto3/__init__.py", line 17, in <module>
2022-09-29 11:21:55.772688 | focal-medium |     from boto3.session import Session
2022-09-29 11:21:55.772695 | focal-medium |   File "/home/ubuntu/src/review.opendev.org/openstack/charm-cinder/.tox/func-target/lib/python3.8/site-packages/boto3/session.py", line 17, in <module>
2022-09-29 11:21:55.772702 | focal-medium |     import botocore.session
2022-09-29 11:21:55.772708 | focal-medium |   File "/home/ubuntu/src/review.opendev.org/openstack/charm-cinder/.tox/func-target/lib/python3.8/site-packages/botocore/session.py", line 26, in <module>
2022-09-29 11:21:55.772715 | focal-medium |     import botocore.client
2022-09-29 11:21:55.772721 | focal-medium |   File "/home/ubuntu/src/review.opendev.org/openstack/charm-cinder/.tox/func-target/lib/python3.8/site-packages/botocore/client.py", line 17, in <module>
2022-09-29 11:21:55.772750 | focal-medium |     from botocore import waiter, xform_name
2022-09-29 11:21:55.772758 | focal-medium |   File "/home/ubuntu/src/review.opendev.org/openstack/charm-cinder/.tox/func-target/lib/python3.8/site-packages/botocore/waiter.py", line 18, in <module>
2022-09-29 11:21:55.772765 | focal-medium |     from botocore.docs.docstring import WaiterDocstring
2022-09-29 11:21:55.772771 | focal-medium |   File "/home/ubuntu/src/review.opendev.org/openstack/charm-cinder/.tox/func-target/lib/python3.8/site-packages/botocore/docs/__init__.py", line 15, in <module>
2022-09-29 11:21:55.772778 | focal-medium |     from botocore.docs.service import ServiceDocumenter
2022-09-29 11:21:55.772785 | focal-medium |   File "/home/ubuntu/src/review.opendev.org/openstack/charm-cinder/.tox/func-target/lib/python3.8/site-packages/botocore/docs/service.py", line 14, in <module>
2022-09-29 11:21:55.772791 | focal-medium |     from botocore.docs.client import ClientDocumenter, ClientExceptionsDocumenter
2022-09-29 11:21:55.772798 | focal-medium |   File "/home/ubuntu/src/review.opendev.org/openstack/charm-cinder/.tox/func-target/lib/python3.8/site-packages/botocore/docs/client.py", line 14, in <module>
2022-09-29 11:21:55.772808 | focal-medium |     from botocore.docs.example import ResponseExampleDocumenter
2022-09-29 11:21:55.772834 | focal-medium |   File "/home/ubuntu/src/review.opendev.org/openstack/charm-cinder/.tox/func-target/lib/python3.8/site-packages/botocore/docs/example.py", line 13, in <module>
2022-09-29 11:21:55.772849 | focal-medium |     from botocore.docs.shape import ShapeDocumenter
2022-09-29 11:21:55.772863 | focal-medium |   File "/home/ubuntu/src/review.opendev.org/openstack/charm-cinder/.tox/func-target/lib/python3.8/site-packages/botocore/docs/shape.py", line 19, in <module>
2022-09-29 11:21:55.772900 | focal-medium |     from botocore.utils import is_json_value_header
2022-09-29 11:21:55.772913 | focal-medium |   File "/home/ubuntu/src/review.opendev.org/openstack/charm-cinder/.tox/func-target/lib/python3.8/site-packages/botocore/utils.py", line 34, in <module>
2022-09-29 11:21:55.772926 | focal-medium |     import botocore.httpsession
2022-09-29 11:21:55.772948 | focal-medium |   File "/home/ubuntu/src/review.opendev.org/openstack/charm-cinder/.tox/func-target/lib/python3.8/site-packages/botocore/httpsession.py", line 46, in <module>
2022-09-29 11:21:55.772964 | focal-medium |     from urllib3.contrib.pyopenssl import (
2022-09-29 11:21:55.772978 | focal-medium |   File "/home/ubuntu/src/review.opendev.org/openstack/charm-cinder/.tox/func-target/lib/python3.8/site-packages/urllib3/contrib/pyopenssl.py", line 50, in <module>
2022-09-29 11:21:55.772991 | focal-medium |     import OpenSSL.SSL
2022-09-29 11:21:55.773003 | focal-medium |   File "/home/ubuntu/src/review.opendev.org/openstack/charm-cinder/.tox/func-target/lib/python3.8/site-packages/OpenSSL/__init__.py", line 8, in <module>
2022-09-29 11:21:55.773016 | focal-medium |     from OpenSSL import SSL, crypto
2022-09-29 11:21:55.773028 | focal-medium |   File "/home/ubuntu/src/review.opendev.org/openstack/charm-cinder/.tox/func-target/lib/python3.8/site-packages/OpenSSL/SSL.py", line 19, in <module>
2022-09-29 11:21:55.773040 | focal-medium |     from OpenSSL.crypto import (
2022-09-29 11:21:55.773052 | focal-medium |   File "/home/ubuntu/src/review.opendev.org/openstack/charm-cinder/.tox/func-target/lib/python3.8/site-packages/OpenSSL/crypto.py", line 3224, in <module>
2022-09-29 11:21:55.773064 | focal-medium |     utils.deprecated(
2022-09-29 11:21:55.773082 | focal-medium | TypeError: deprecated() got an unexpected keyword argument 'name'
```

Reply:
pyopenssl is an optional dependency of urllib3 - https://github.com/urllib3/urllib3/blob/main/src/urllib3/contrib/pyopenssl.py#L168 , so you are right, at install time won't be pulled in as an indirect dependency of boto3.